File: cpp/0113_pathSum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// 1. DFS,每层累加number,用一个数组来记录路径，用depth记录深度以及控制是否erase数组最后一个元素（如果数组的长度大于深度则erase）
// 2. 如果number==target && 是叶子节点，记录
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> res;
        vector<int> arr;
        traverse(root,0,targetSum,res,arr,1);
        return res;
    }
    void traverse(TreeNode* root,int sum, int targetSum, vector<vector<int>> &res, vector<int> &arr, int depth)
    {
        if(root==nullptr)
            return;  
        
        while(arr.size()>=depth)
            arr.erase(arr.end()-1);
        arr.push_back(root->val);
        
        if(root->right==nullptr && root->left==nullptr && sum+root->val == targetSum)
            res.push_back(arr);
        traverse(root->left,sum+root->val,targetSum,res,arr,depth+1);
        traverse(root->right,sum+root->val,targetSum,res,arr,depth+1);
    }
};

File: java/0113_pathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// 小心java的数组都是对象，应该copy数组
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> arr = new ArrayList<Integer>();
        traverse(root,targetSum,0,1,res,arr);
        return res;
    }
    public void traverse(TreeNode root,int targetSum,int sum, int depth,List<List<Integer>> res, List<Integer> arr)
    {
        if(root==null)
            return;
        while(arr.size()>=depth)
            arr.remove(arr.size()-1);
        arr.add(root.val);
        if(root.right==null && root.left==null && root.val+sum==targetSum)
            res.add(new ArrayList<Integer>(arr));
        traverse(root.left,targetSum,sum+root.val,depth+1,res,arr);
        traverse(root.right,targetSum,sum+root.val,depth+1,res,arr);
    }
}
